package/upgrade/upgrade_manifests.sh

```diff
@@ -451,6 +452,76 @@ EOF
   wait_rollout cattle-monitoring-system deployment rancher-monitoring-operator
 }
 
+#upgrade only monitoring_crd
+upgrade_managedchart_monitoring_crd() {
+  local nm=rancher-monitoring-crd
+  echo "Upgrading Managedchart $nm to $REPO_MONITORING_CHART_VERSION"
+
+  local pre_version=$(kubectl get managedcharts.management.cattle.io "$nm" -n fleet-local -o=jsonpath='{.spec.version}')
+  if [ "$pre_version" = "$REPO_MONITORING_CHART_VERSION" ]; then
+    echo "the $nm has already been target version $REPO_MONITORING_CHART_VERSION, nothing to upgrade"
+    pause_managed_chart "$nm" "false"
+    return 0
+  fi
+
+  local pre_generation=$(kubectl get managedcharts.management.cattle.io "$nm" -n fleet-local -o=jsonpath='{.status.observedGeneration}')
+
+  mkdir -p $UPGRADE_TMP_DIR/monitoring
+  cd $UPGRADE_TMP_DIR/monitoring
+
+  cat >"$nm".yaml <<EOF
+spec:
+  version: $REPO_MONITORING_CHART_VERSION
+EOF
+
+  kubectl patch managedcharts.management.cattle.io "$nm" -n fleet-local --patch-file ./"$nm".yaml --type merge
+
+  pause_managed_chart "$nm" "false"
+
+  wait_managed_chart fleet-local "$nm" $REPO_MONITORING_CHART_VERSION $pre_generation ready
+}
+
+upgrade_monitoring() {
+  if [[ "$UPGRADE_PREVIOUS_VERSION" > "v1.1.1" ]] && [[ "$UPGRADE_PREVIOUS_VERSION" < "v1.2.0" ]]; then
+    upgrade_managedchart_monitoring_crd
+    convert_monitoring_to_addon
+  elif [[ "$UPGRADE_PREVIOUS_VERSION" = "v1.2.0" ]] || [[ "$UPGRADE_PREVIOUS_VERSION" > "v1.2.0" ]]; then
+    # from v1.2.0, only crd here, rancher-monitoring is upgraded in addons
+    upgrade_managedchart_monitoring_crd
+  else
+    # legacy, upgrade both
+    upgrade_managedchart_monitoring_and_crd
+  fi
+}
+
+upgrade_managedchart_logging_crd() {
+  nm=rancher-logging-crd
+  echo "Upgrading Managedchart $nm to $REPO_LOGGING_CHART_VERSION"
+
+  pre_version=$(kubectl get managedcharts.management.cattle.io "$nm" -n fleet-local -o=jsonpath='{.spec.version}')
+  if [ "$pre_version" = "$REPO_LOGGING_CHART_VERSION" ]; then
+    echo "the $nm has already been target version $REPO_LOGGING_CHART_VERSION, nothing to upgrade"
+    pause_managed_chart "$nm" "false"
```

Comment:
The conversion works excellently when upgrading from v1.1.2 to master-build. But seeing this error when upgrade to master build from master build again (`previous_version` is `master-eeaaf82a-head`)
```
hvst-upgrade-b5rn5-apply-manifests-m6d25 Current version: 0.0.0-master-eeaaf82a, Current state: null, Current generation: 10
hvst-upgrade-b5rn5-apply-manifests-m6d25 Waiting for ManagedChart fleet-local/harvester-crd from generation 7
hvst-upgrade-b5rn5-apply-manifests-m6d25 Target version: 0.0.0-master-eeaaf82a, Target state: ready
hvst-upgrade-b5rn5-apply-manifests-m6d25 Current version: 0.0.0-master-eeaaf82a, Current state: null, Current generation: 8
hvst-upgrade-b5rn5-apply-manifests-m6d25 Waiting for KubeVirt to upgraded to 0.54.0-150400.3.16.1...
hvst-upgrade-b5rn5-apply-manifests-m6d25 Waiting for LH settling down...
hvst-upgrade-b5rn5-apply-manifests-m6d25 Waiting for longhorn-manager to be upgraded...
hvst-upgrade-b5rn5-apply-manifests-m6d25 Checking instance-manager-r pod on node node1...
hvst-upgrade-b5rn5-apply-manifests-m6d25 Checking instance-manager-r pod on node node1 OK.
hvst-upgrade-b5rn5-apply-manifests-m6d25 Checking instance-manager-r pod on node node2...
hvst-upgrade-b5rn5-apply-manifests-m6d25 Checking instance-manager-r pod on node node2 OK.
hvst-upgrade-b5rn5-apply-manifests-m6d25 Checking instance-manager-r pod on node node3...
hvst-upgrade-b5rn5-apply-manifests-m6d25 Checking instance-manager-r pod on node node3 OK.
hvst-upgrade-b5rn5-apply-manifests-m6d25 Upgrading Monitoring
hvst-upgrade-b5rn5-apply-manifests-m6d25 Error from server (NotFound): managedcharts.management.cattle.io "rancher-monitoring" not found
Stream closed EOF for harvester-system/hvst-upgrade-b5rn5-apply-manifests-m6d25 (apply)
```

It should be OK if the QA uses RC release to test the upgrade from RC to RC. Can we also handle this case to ease the test of master-build.

Reply:
I will check how to identify the version like `0.0.0-master-eeaaf82a`, are all master-head versions  like `0.0.0-*` ?

Reply:
Added a new commit to handle the none-formal release, and tested from one master-head to another, looks OK.

@bk201 please re-check, thanks.

```
...
Upgrading Managedchart rancher-monitoring-crd to 102.0.0+up40.1.2
the rancher-monitoring-crd has already been target version 102.0.0+up40.1.2, nothing to upgrade
managedchart.management.cattle.io/rancher-monitoring-crd patched

Monitoring: start to convert monitoring from managedchart to addon
monitoring has already been addon

The current version is master-5557c692-head, will check Logging Event Audit upgrade manifest option
Upgrading Managedchart rancher-logging-crd to 102.0.0+up3.17.10
the rancher-logging-crd has already been target version 102.0.0+up3.17.10, nothing to upgrade
managedchart.management.cattle.io/rancher-logging-crd patched

Logging Event Audit: start to convert logging from managedchart to addon
logging has already been addon

Applying extra manifests
..
addon.harvesterhci.io/vm-import-controller patched
addon.harvesterhci.io/pcidevices-controller patched (no change)

upgrade addon rancher_monitoring
try to patch addon rancher-monitoring in cattle-monitoring-system 102.0.0+up40.1.2
addon has already been 102.0.0+up40.1.2, nothing to do

upgrade addon rancher_logging
try to patch addon rancher-logging in cattle-logging-system 102.0.0+up3.17.10
addon has already been 102.0.0+up3.17.10, nothing to do

```